Cql/Cql.CqlToElm/TypeConverter.cs

```diff
@@ -0,0 +1,426 @@
+                new Interval { low = As(expression, type.pointType), high = As(expression, type.pointType) }.WithResultType(type);
+
+            Expression FromList(Expression list, TypeSpecifier to) =>
+                As(new SingletonFrom { operand = list }.WithResultType(to), to);
+
+            Expression FromInterval(Expression interval, TypeSpecifier to) =>
+                As(new Start { operand = interval }.WithResultType(to), to);
+
+            Expression ToList(Expression expression, ListTypeSpecifier to) =>
+                new ToList { operand = As(expression, to.elementType) }.WithResultType(to);
+        }
+
+        /// <summary>
+        /// Computes the cost of conversion according to the specification's conversion precendence rules.
+        /// </summary>
+        /// <seealso href="https://cql.hl7.org/03-developersguide.html#conversion-precedence"/>
+        public ConversionCost GetConversionCost(TypeSpecifier from, TypeSpecifier to)
+        {
+            if (IsExactMatch(from, to))
+                return ConversionCost.ExactMatch;
+            else if (IsSubtype(from, to))
+                return ConversionCost.Subtype;
+            else if (IsCompatible(from, to))
+                return ConversionCost.Compatible;
+            else if (CanBeCast(from, to))
+                return ConversionCost.Cast;
+            else if (HasImplicitConversion(from, to))
+            {
+                if (IsSimpleType(to))
+                    return ConversionCost.ImplicitToSimpleType;
+                else if (IsClassType(to))
+                    return ConversionCost.ImplicitToClassType;
+                else return ConversionCost.Incompatible;
+            }
+            else if ((Options.EnableIntervalPromotion ?? false)
+                && to is IntervalTypeSpecifier promotableInterval
+                && CanBePromoted(from, promotableInterval))
+                return ConversionCost.IntervalPromotion;
+            else if ((Options.EnableListDemotion ?? false)
+                && from is ListTypeSpecifier listType && CanBeDemoted(listType, to))
+                return ConversionCost.ListDemotion;
+            else if ((Options.EnableIntervalDemotion ?? false)
+                && from is IntervalTypeSpecifier demotableInterval && CanBeDemoted(demotableInterval, to))
+                return ConversionCost.IntervalDemotion;
+            else if ((Options.EnableListPromotion ?? false)
+                && to is ListTypeSpecifier promotableListType && CanBePromoted(from, promotableListType))
+                return ConversionCost.ListPromotion;
+            else
+                return ConversionCost.Incompatible;
+        }
+
+        internal string IncompatibilityMessage(TypeSpecifier from, TypeSpecifier to) =>
+            $"Expression of type '{from}' cannot be cast as a value of type '{to}'.";
+
+        // All types exactly match Any
+        internal virtual bool IsExactMatch(TypeSpecifier from, TypeSpecifier to) =>
```

Comment:
I felt like we may encounter scenarios where the only way to support certain CQL files was to intentionally subvert a small part of the coercion algorithm so we could match a buggy implementation's behavior, while preserving the rest of the behavior.  But since these methods are internal we can just make them virtual when and if that happens, or we can just create an interface for this service and inject that interface instead of the implementation.  I'll remove it.